### PR TITLE
Add AMQP Broker

### PR DESCRIPTION
**Pull request by usererrorllc**
Please review MR to add AMQP Broker. Let me know if any questions.

### CODE REVIEW

File: composer.json
Comment (smuuf) [inline composer.json]: Not sure why `symfony/polyfill-iconv` dependency is necessary...? 🤔 Is it really? I don't see any `*iconv*` functions used anywhere _(but I may be blind)_.

File: src/Drivers/IAmqpDriver.php
Comment (smuuf) [inline src/Drivers/IAmqpDriver.php]: I see that `PhpAmqpLibAmqpDriver` implementation of `IAmqpDriver` does return `true,` but I don't see any reason for it to do that, as the return value is not used anywhere.

 Fox that reason I think both the `IAmqpDriver` interface and `PhpAmqpLibAmqpDriver` should have `void` return type for this method.

File: src/Drivers/PhpAmqpLibAmqpDriver.php
Comment (smuuf) [inline src/Drivers/PhpAmqpLibAmqpDriver.php]: Same as with the `PhpAmqpLibAmqpDriver::publish()` method. The return value is not used anywhere, yet it needlessly returns `true` (and also the return type is missing).

This should also be of `void` return type.

File: src/Drivers/PhpAmqpLibAmqpDriver.php
Comment (smuuf) [inline src/Drivers/PhpAmqpLibAmqpDriver.php]: Same as above. Should have return type `void` and shouldn't unnecessarily return `true`, which isn't used anywhere anyway.

File: src/Drivers/PhpAmqpLibAmqpDriver.php
Comment (smuuf) [inline src/Drivers/PhpAmqpLibAmqpDriver.php]: This method doesn't seem to be used anywhere...? It seems like this is meant to be used in situations where AMQP/RabbitMQ should act as a result backend - but that's doesn't seem to be relevant for this PR _(and, again, this method is unused)_.

Consider removing this method _(from both here and from the `IAmqpDriver` interface, along with the `consumeResult` method called from this one)_.

File: src/Drivers/PhpAmqpLibAmqpDriver.php
Comment (smuuf) [inline src/Drivers/PhpAmqpLibAmqpDriver.php]: See comment for the method above.

**Review by smuuf — CHANGES_REQUESTED** [review]
Thank you for the PR! The AMQP/RabbitMQ broker support will be a welcomed addition. 🍀

However I have some issues with the present changes. Some of those are written as comments in specific places and some others are here, down below:

## Code style
The celery-for-php project **uses tab-based indentation** for PHP files and not space-based. Then there are other inconsistencies like `}else {` or ˙if(! empty($routingKey) ){˙ _(in contrast to `if ($e->getCode() !== 404) {` - which is fine)_.

To help you with this **I implemented code style linting yesterday via [php-cs-fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer) with the ability to fix such issues.** It's now merged in `master` and [described here](https://github.com/smuuf/celery-for-php/pull/3) _(this unfortunately introduced some minor conflicts with Composer stuff changed in this PR of yours, but I'm sure you'll be able to resolve them just fine)_.

Please run `./bin/code-style.sh` to see code style issues - or just use `./bin/code-style.sh --fix` to directly fix them _(maybe you'll need to `chmod +x ./bin/code-style.sh` it, I'm not sure if git committed it with correct execution flags)_; then commit the fixed changes and push then into your PR branch.

Unfortunately `php-cs-fixer` is (apparently) unable to deal with some additional stuff - which I pointed out in separate comments.

## Testing
The whole idea of having a testing environment done via Docker was to have a environment-agnostic framework where the tests could be executed. 

The thing is: I don't have RabbitMQ installed on my local env and I don't want to have it installed just to run celery-for-php's tests. That's why the ***tests should be executed in Docker*** - and that's why **the dot-env stuff is IMO unnecessary**.- I don't _(at least currently)_ see the need to have to create a `.env` file just to run tests.

Additionally, your `CELERYFORPHP_REDIS_HOST=127.0.0.1` is different from the original `[::1]` Celery host IP used in tests. This (ipv6) addressing was working for tests running in GitHub Actions (with Docker), tests running with my combined WSL1-WSL2 local environment and my seconds pure-WSL2 environment. You changes broke at least the latter two 🙂 

Also, as far as I can see, the AMQP broker is not covered properly with integration tests _(your tests do cover that **some** messages are being sent into AMQP/RabbitMQ, but then we don't have any way of knowing whether Celery would in fact receive these messages and/or interpret them as tasks)_.

ℹ️ I understand that it's not very clear how tests for a completely new broker should be integrated into the way the tests currently are - but ***this is something I'd be gladly willing to rework myself***. I would just like you to give me a bit of a cleaner slate:

1. Keep the `Brokers.AmqpBroker.phpt`, it's good as it is.
2. Remove ˙vlucas/phpdotenv˙ Composer dependency.
3. Revert the changes in `CeleryFactory`.

---

Other than these I'm happy with your contribution and I'm looking forward to merging it, it just needs some polishing🌻 Thanks!

File: composer.json
Comment (usererrorllc) [reply composer.json]: This was a sub-dependency from dotenv -- since we're yanking dotenv out I'll yank that out too.

**Comment by usererrorllc** [comment]
Thanks for your review -- if I missed anything just let me know and I'll get the updates done. :)

**Comment by smuuf** [comment]
I've updated the testing suite to accommodate for multiple brokers vs (potentially) multiple result backends. Both work for local tests and for GHA tests. 

Also I've - hopefully correctly - fixed a bug in declaring exchange for Celery pidbox, where Celery was declaring it as `durable=False`, but your broker implementation was trying to to declare it with `durable: true`, which was then causing `AMQPProtocolChannelException: PRECONDITION_FAILED - inequivalent arg 'durable' for exchange 'celery.pidbox' in vhost '/': received 'true' but current is 'false'` - so I took the liberty of correcting that one.

Alas it seems that phpstan complains about some stuff in the code of `PhpAmqpLibAmqpDriver` class _(it was complaining even before my latest commit into your PR, but I noticed it just now)_. Could you take a look at it? :)

The unused `$channel` property is an easy fix. The issue with `Dead catch - PhpAmqpLib\Exception\AMQPProtocolChannelException is never thrown in the try block.` also doesn't seem like something complicated, but I'd feel better if you'd take a look and fix it.

**Comment by usererrorllc** [comment]
@smuuf 

Yeah that leftover $channel var was from the fetchResult method that we removed. phpstan is complaining about these exceptions not firing, but they are thrown by the parent lib. php-amqplib does not have an exception-free way to check if a queue/exchange exists that I know of (it throws the caught exception). Perhaps there is a way to hint to phpstan in a better way to resolve this -- but I added a baseline with those 3 errors and it's now passing.

I'll leave it up to you if that is acceptable or not! If you have a better solution, I appreciate your help there :)

Thanks!

**Comment by usererrorllc** [comment]
By the way, decreasing the phpstan --level to 3 will also pass without the baseline.

**Comment by smuuf** [comment]
1. Lowering phpstan level is just hiding the problem, not solving it.
2. Baseline is not the way this should be handled. The idea of having a baseline is to make transition to phpstan easier in projects that haven't been using it. Then things should continuously disappear from baseline rather than appear :)
3. I did some testing around the expected exception and discovered:
    1. Executing `$ch->exchange_delete("nonexistent_deleteExchange");` does _not_ seem to throw any exception.
    2. Executing `$ch->queue_delete("nonexistent_deleteQueue");` does _not_ seem to throw any exception.
    3. Executing `$ch->queue_purge("nonexistent_purgeQueue");` _does_ indeed throw exception: `PhpAmqpLib\Exception\AMQPProtocolChannelException: NOT_FOUND - no queue 'nonexistent_purgeQueue' in vhost '/'`

Ad `3.` it seems to me that both `exchange_delete` nor `queue_delete` do not throw _any_ exception (let alone `AMQPProtocolChannelException`). That may be because I was doing something wrong when checking those, or it might be a bug in php-amqplib (?).

So, these two cases I'm not sure the `try...catch` should be even there _(I'd even say it shouldn't be there...)_. But you may know better, so I leave it up to you. If you decide the `try...catch` should really be there, then I'd do it the same way as I describe below.

The `$ch->queue_purge("nonexistent_purgeQueue")` call does throw and the `try...catch` is in that case justified. PHPStan still complains that it's a "dead catch", very probably due to the IMHO highly-magic implementation of php-amqplib's internals.

So, in this case - instead of lowering phpstan level or creating a baseline - you should just [tell phpstan not to check that specific place](https://phpstan.org/user-guide/ignoring-errors) via a `// @phpstan-ignore-line` comment at that specific place:

```php
try {
	$ch->queue_purge($queueName);
} catch (AMQPProtocolChannelException $e) { // @phpstan-ignore-line
	if ($e->getCode() !== 404) {
		throw $e;
	}
}
```

**Comment by usererrorllc** [comment]
I replaced my latest commit adjusting to your response.

1. Although I've never used phpstan before I do understand that much.
2. I hunted around for exactly what you just shared with me, thanks! I updated with an ignore line -- much cleaner.
3. I think I actually only ran into the exception with purge_queue when making updates to get tests to work. Thanks for the time testing that.

We should be ready to rock now!

**Comment by smuuf** [comment]
Great! 

I'd like to do some additional cosmetic changes before I merge it into master. Would it be ok if I'd do a forced push into your branch? (I'm not sure GitHub will allow me, but if yes, then I'd also like to clean up the commit graph a bit).

**Comment by usererrorllc** [comment]
Yeah that’s cool with me!

**Review by smuuf — APPROVED** [review]
Thanks! 🌻

**Pull request merged**
